sync/commit.py

```diff
@@ -134,10 +136,38 @@ def create(cls, repo, msg, metadata, author=None, amend=False):
 
     @staticmethod
     def make_commit_msg(msg, metadata):
+        if metadata is None:
+            metadata = {}
+
+        msg_lines = msg.splitlines()
+        metadata_seen = []
+        keep_lines = None
+        for i, line in enumerate(reversed(msg_lines)):
+            if line.strip() == "":
+                continue
+            m = METADATA_RE.match(line)
+            if m:
+                metadata_seen.append(m.groups()[0])
+            else:
+                keep_lines = len(msg_lines) - i
+                break
+
+        msg_lines = msg_lines[:max(keep_lines, 1)]
+
         if metadata:
-            metadata_str = "\n".join("%s: %s" % item for item in sorted(metadata.items()))
-            msg = "%s\n%s" % (msg, metadata_str)
-        return msg
+            msg_lines.append("")
+
+        metadata_added = set()
```

Comment:
I think this can be simplified. Judging by the way `test_metadata_update` is written, we always replace the commit message metadata entirely with what's in `metadata`, so it would work to just write everything in `metadata` right after the "kept lines"? That is, I don't think we actually need to keep track of `metadata_seen` unless it's important to maintain the same order as before...